packages/react-reconciler/src/ReactFiberWorkLoop.js

```diff
@@ -727,6 +729,7 @@ export function scheduleUpdateOnFiber(
   }
 
   // Mark that the root has a pending update.
+  markRootWithPopState(root, lane);
```

Comment:
scheduleUpdateOnFiber is incredibly hot. The scenario to keep in mind with these is something like a Flux framework or Relay calling this thousands of times to schedule updates. It's already doing too much and this is adding a check that has to go through unoptimized paths to check window.event.

Is there a more suitable place you can place this where we already check for a transition? Such as in requestUpdateLane where we know if it's a transition and then we can check this.

I think we need to do a larger refactor this whole flow because it's too spread out and setState is way too costly.

We should be able to get this to just set the dirty flags on the parent nodes given a priority bitmask already computed for the environment + check if we've scheduled a microtask. Then inside the microtask we can check if we need to schedule postTask or rAF or flush it right away given window.event. With this strategy we can batch all this work that happens for each setState call to just happen once per task. The new model is also vastly simplified so some of the old considerations don't apply.

Reply:
Makes sense! I'm starting it on a new branch, and I have several questions. Just wondering if you already have ideas about how to deal with `queueMicrotask` in tests? With the change, some tests need to assert after the new microtask. I think we need to wrap updates in `act` if they haven't or we would need to move some test assertions in a queueMicrotask callback?

> I think we need to do a larger refactor this whole flow because it's too spread out and setState is way too costly.

Totally agree! I found it hard to connect pieces. Just to clarify, is the larger refactor a superset of what needed to change about eagerly calling microtask?


> We should be able to get this to just set the dirty flags on the parent nodes given a priority bitmask already computed for the environment + check if we've scheduled a microtask.

Is the dirty flags for things related to attemptEagerTransition?